include/openssl/hpke.h

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* APIs and data structures for HPKE (RFC9180)  */
+#ifndef OSSL_HPKE_H
+# define OSSL_HPKE_H
+# pragma once
+
+# include <openssl/obj_mac.h>
+# include <openssl/types.h>
+
+/* HPKE modes */
+# define OSSL_HPKE_MODE_BASE              0 /* Base mode  */
+# define OSSL_HPKE_MODE_PSK               1 /* Pre-shared key mode */
+# define OSSL_HPKE_MODE_AUTH              2 /* Authenticated mode */
+# define OSSL_HPKE_MODE_PSKAUTH           3 /* PSK+authenticated mode */
+
+/*
+ * The (16bit) HPKE algorithn ID IANA codepoints
+ * If/when new IANA codepoints are added there are tables in
+ * crypto/hpke/hpke_util.c that must also be updated.
+ */
+# define OSSL_HPKE_KEM_ID_RESERVED         0x0000 /* not used */
+# define OSSL_HPKE_KEM_ID_P256             0x0010 /* NIST P-256 */
+# define OSSL_HPKE_KEM_ID_P384             0x0011 /* NIST P-384 */
+# define OSSL_HPKE_KEM_ID_P521             0x0012 /* NIST P-521 */
+# define OSSL_HPKE_KEM_ID_X25519           0x0020 /* Curve25519 */
+# define OSSL_HPKE_KEM_ID_X448             0x0021 /* Curve448 */
+
+# define OSSL_HPKE_KDF_ID_RESERVED         0x0000 /* not used */
+# define OSSL_HPKE_KDF_ID_HKDF_SHA256      0x0001 /* HKDF-SHA256 */
+# define OSSL_HPKE_KDF_ID_HKDF_SHA384      0x0002 /* HKDF-SHA384 */
+# define OSSL_HPKE_KDF_ID_HKDF_SHA512      0x0003 /* HKDF-SHA512 */
+
+# define OSSL_HPKE_AEAD_ID_RESERVED        0x0000 /* not used */
+# define OSSL_HPKE_AEAD_ID_AES_GCM_128     0x0001 /* AES-GCM-128 */
+# define OSSL_HPKE_AEAD_ID_AES_GCM_256     0x0002 /* AES-GCM-256 */
+# define OSSL_HPKE_AEAD_ID_CHACHA_POLY1305 0x0003 /* Chacha20-Poly1305 */
+# define OSSL_HPKE_AEAD_ID_EXPORTONLY      0xFFFF /* export-only fake ID */
+
+/* strings for suite components - ideally these'd be defined elsewhere */
+# define OSSL_HPKE_KEMSTR_P256        "P-256"              /* KEM id 0x10 */
```

Comment:
I fthese are not meant to be public, put them in the internal header.

Reply:
Yeah fair comment. I think they're better public here as they make up the good inputs to ``OSSL_HPKE_str2suite()``. 
I also have a command line tool (for ECH key gen) that prints these out when run with ``-help``.

I also changed to allow the special "exporter" AEAD in ``OSSL_HPKE_str2suite()`` too and included the literal strings here (e.g. ``"X25519"`` rather than ``SN_X25519``) as I've never found the full set of strings in other headers that'd make sense to use. If they do exist somewhere else I'd be happy to use e.g. ``SN_X25519`` but I guess better to be consistent and there's no real chance those strings will change over time is there?